ci/generate_docs.py

```diff
@@ -0,0 +1,219 @@
+        if not os.path.exists(self.GITLAB_CI_FILE):
+            raise RuntimeError(
+                f"Main Gitlab CI file under {self.GITLAB_CI_FILE} does not exist!"
+            )
+
+        with open(self.GITLAB_CI_FILE, "r") as f:
+            gitlab_file_content = yaml.safe_load(f)
+
+        all_ci_files = gitlab_file_content["include"]
+
+        for file in all_ci_files:
+            if not os.path.isfile(file):
+                raise RuntimeError(f"File {file} does not exist!")
+
+        return all_ci_files
+
+    @staticmethod
+    def get_overall_description_from_file(file: str) -> List[str]:
+        """Looking for comments at the very beginning of the file."""
+        description_lines: List[str] = []
+        with open(file, "r") as f:
+            for line in f:
+                if line.startswith("#"):
+                    # Taking just the text - no hashes, no whitespace
+                    description_lines.append(line.strip("# \n"))
+                else:
+                    break
+
+        return description_lines
+
+    def get_jobs_from_file(self, file: str) -> Dict[str, Dict[str, Any]]:
+        """Extract all jobs and their details from a certain file."""
+        all_jobs: Dict[str, Dict[str, Any]] = OrderedDict()
+
+        # Getting the parsed content of the file, so we can get the script array
+        with open(file, "r") as f:
+            gitlab_file_content = yaml.safe_load(f)
+
+        # Taking all the comments above a non-indented non-comment, which is
+        # always a job definition, unless defined in NOT_JOBS
+        with open(file, "r") as f:
+            comment_buffer: List[str] = []
+            for index, line in enumerate(f):
+                if line.startswith("#"):
+                    # Taking just the text - no hashes, no whitespace
+                    comment_buffer.append(line.strip("# \n"))
+                else:
+                    # regex: first character of a line is a word-character or a dot
+                    if re.search(r"\A[\w\.]", line) and not any(
+                        [line.startswith(not_job) for not_job in self.NOT_JOBS]
+                    ):
+                        job_name = line.rstrip(":\n")
+                        if job_name in gitlab_file_content:
+                            all_jobs[job_name] = {
+                                "description": comment_buffer,
+                                "line_no": index + 1,
+                                "script": gitlab_file_content[job_name].get(
+                                    "script",
+                                    ["No script defined, probably extends another job"],
+                                ),
+                            }
+                    comment_buffer = []
+
+        return all_jobs
+
+    def save_docs_into_file(self) -> None:
+        """Dump all the information into a documentation file."""
+        with open(self.DOC_FILE, "w") as doc_file:
```

Comment:
In my experience using templates ([mako](https://www.makotemplates.org/) in our case) for generating files tends to result in more maintainable code. But you can probably rewrite it when a change will be needed for the first time.